pubtools/_quay/command_executor.py

```diff
@@ -245,3 +258,165 @@ def _run_cmd(self, cmd, err_msg=None, tolerate_err=False, stdin=None):
 
         return out_text, err_text
+
+
+class ContainerExecutor(Executor):
+    """Run commands in a Docker container."""
+
+    def __init__(
+        self,
+        image,
+        base_url="unix://var/run/docker.sock",
+        timeout=None,
+        verify_tls=False,
+        cert_path=None,
+    ):
+        """
+        Initialize.
+
+        Args:
+            image (str):
+                Path to an image which will be used for performing skopeo operations. Must be
+                downloadable by Docker.
+            base_url (str):
+                Base URL of the Docker client.
+            timeout (int):
+                Default timeout for API calls, in seconds.
+            verify_tls (bool):
+                Whether to use TLS verification.
+            cert_path (str|None):
+                Custom path to TLS certificates. If not specified, '~/.docker' is used.
+        """
+        self.image = image
+
+        kwargs = {}
+        kwargs["base_url"] = base_url
+        if timeout:
+            kwargs["timeout"] = timeout
+        if verify_tls:
+            kwargs["base_url"] = kwargs["base_url"].replace("tcp://", "https://")
+            cert_path = cert_path or os.path.join(os.path.expanduser("~"), ".docker")
+            if os.path.isdir(cert_path):
+                kwargs["tls"] = docker.tls.TLSConfig(
+                    client_cert=(
+                        os.path.join(cert_path, "cert.pem"),
+                        os.path.join(cert_path, "key.pem"),
+                    ),
+                    verify=os.path.join(cert_path, "ca.pem"),
+                )
+
+        self.client = docker.APIClient(**kwargs)
+        repo, tag = self.image.split(":", 1)
+        self.client.pull(repo, tag=tag)
+        self.container = self.client.create_container(self.image, detach=True, tty=True)
+        self.client.start(self.container["Id"])
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        """Cleanup the container when used as a context manager."""
+        self.client.remove_container(self.container["Id"], force=True)
+
+    def _run_cmd(self, cmd, err_msg=None, tolerate_err=False, stdin=None):
+        """
+        Run a command locally.
+
+        Args:
+            cmd (str):
+                Shell command to be executed.
+            error_msg (str):
+                Error message written when the command fails.
+            tolerate_err (bool):
+                Whether to tolerate a failed command.
+            stdin (None|str):
+                This parameter exists only for compatibility with parent class. Sending input
+                to containers is not supported.
+
+        Returns (str, str):
+            Tuple of stdout and stderr generated by the command.
+        """
+        err_msg = err_msg or "An error has occured when executing a command."
+        cmd_exec = self.client.exec_create(self.container["Id"], cmd)
+        stdout, stderr = self.client.exec_start(cmd_exec["Id"], demux=True)
```

Comment:
exec_api installed by python-docker-py-1.7.2-1.el6eng.noarch doesn't have argument 'demux'.

Reply:
Good point, I removed the parameter. The unfortunate side effect is that now we can't differentiate stdout and stderr. To maintain consistency with other classes, I duplicated the output to be returned as tuple. I think this change shouldn't affect any of the current use-cases.